### PR TITLE
feat(tracing): allow sampling based on tags

### PR DESCRIPTION
**Pull request by emmettbutler**
This change implements tag-based trace sampling by adding the `tags` keyword parameter to `SamplingRule.__init__`. It does not change when `SamplingRule`s are evaluated.

Checking tags during sampling has a negative impact on performance, which we're willing to accept for the product value associated with this feature.

## Checklist

- [x] benchmark updates
- [x] DD_TRACE_SAMPLING_RULES envvar support for tags
- [x] envvar tests
- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

File: tests/integration/test_sampling.py [hunk 1]
```diff
@@ -136,3 +137,44 @@ def test_extended_sampling_resource(writer, tracer):
...
+def test_extended_sampling_tags(writer, tracer):
+    sampler = DatadogSampler(rules=[SamplingRule(0, tags=TAGS)])
+    tracer.configure(sampler=sampler, writer=writer)
+    tracer._tags = TAGS
+    tracer.trace("should_not_send").finish()
+    tracer._tags = {"banana": "True"}
+    tracer.trace("should_send").finish()
+
+
+@snapshot_parametrized_with_writers
```
Comment (ZStriker19) [inline tests/integration/test_sampling.py:152]: should we test `?` as well while we're at it?

**Review by ZStriker19 — CHANGES_REQUESTED** [review]
Looks good, just a couple questions and a recommendation.

File: tests/integration/test_sampling.py (hunk 1 above)
Comment (emmettbutler) [reply tests/integration/test_sampling.py:152]: sure!

File: ddtrace/sampling_rule.py [hunk 2]
```diff
@@ -131,8 +141,29 @@ def matches(self, span):
         :returns: Whether this span matches or not
         :rtype: :obj:`bool`
         """
-        # self._matches exists to maintain legacy pattern values such as regex and functions
-        return self._matches((span.service, span.name, span.resource))
+        glob_match = self.glob_matches(span)
+        return glob_match and self._matches((span.service, span.name, span.resource))
```
[the remark this answers is not in the document]
Comment (emmettbutler) [reply ddtrace/sampling_rule.py:145]: `_matches` only returns true when service, name, and resource match. This change essentially adds `glob_matches` to the preexisting `and` expression.

[the remark this answers is not in the document]
Comment (ZStriker19) [reply ddtrace/sampling_rule.py:145]: Ah I see got it!

File: ddtrace/sampling_rule.py [hunk 3]
```diff
@@ -131,8 +141,29 @@ def matches(self, span):
...
+
+        tag_match = False
+        for tag_key in self._tag_value_matchers.keys():
+            value = tags.get(tag_key)
+            if value is not None:
+                tag_match = self._tag_value_matchers[tag_key].match(value)
+            else:
+                # if we don't match with all specified tags for a rule, it's not a match
+                return False
+        return tag_match
```
Comment (ZStriker19) [inline ddtrace/sampling_rule.py:166]: Should we change the __eq__ method and __repr__ method?

**Review by ZStriker19 — DISMISSED** [review]

**Review by ZStriker19 — APPROVED** [review]

**Pull request merged**